src/concurrent_block_queue.h

```diff
@@ -0,0 +1,160 @@
+#ifndef CONCURRENT_BLOCK_QUEUE_H
+#define CONCURRENT_BLOCK_QUEUE_H
+
+#include <condition_variable>
+#include <cstddef>
+#include <memory>
+#include <mutex>
+#include <optional>
+#include <utility>
+#include <vector>
+
+#include "Queue.h"
+
+namespace Utils {
+
+  template <typename T, size_t BLOCK_SIZE = 512> 
+    requires ( std::copyable<T> || std::movable<T> )
+    class ConcurrentBlockQueue {
+
+    struct Node {
+      std::vector<T> data = {};
+      std::unique_ptr<Node> next = nullptr;
+
+      Node( ) { data.reserve(BLOCK_SIZE); }
+      ~Node( ) = default;
+      Node( Node&& ) = default;
+      Node& operator=( Node&& ) = default;
+
+      Node( const Node& ) = delete;
+      Node& operator=( const Node& ) = delete;
+    };
+
+    struct Head{
+      std::unique_ptr<Node> head_block = nullptr;
+      size_t block_offset = 0;
+      std::mutex lock;
+
+      T pop_data( ) {
+        auto retval = head_block->data[block_offset]; 
```

Comment:
you could use std::move here to avoid unnecessary copy.

Reply:
https://github.com/ashwinravianandan/ThreadPoolLibrary/pull/2